csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -284,21 +286,25 @@ PersistentBufferStorageParams getPersistentBufferStorageParams(
         InnerOuterPersistentKernelScheduler::threads_per_block_min,
         InnerOuterPersistentKernelScheduler::threads_per_block_max,
         dev_prop->warpSize);
-    buffer_params.smem_buffer_size += tv_buffer_size_smem;
+    smem_buffer_size += tv_buffer_size_smem;
 
     // The first-i buffers to are moved from register to shared memory
     // If both the register buffer size and shared memory buffer size are within
     // the allowable limit, we found a good configuration. Record the number of
-    // buffers to be moved to shared memory and break the loop.
-    if (buffer_params.regs_buffer_size <= available_regs &&
-        buffer_params.smem_buffer_size <= available_smem) {
+    // buffers to be moved to shared memory and continue checking if we can move
+    // more. Moving more buffers to shared memory reduces the usage of
+    // registers, the comipler can uses more register to optimize the kernel for
+    // better instruction level parallelism and leads to better performance.
+    if (regs_buffer_size <= available_regs &&
```

Comment:
It is still necessary. We begin with `{regs_buffer = N, smem_buffer = 0}`. In each iteration, one buffer is moved from registers to shared memory, resulting in `{regs_buffer = N-x, smem_buffer = x}`. We must check the conditions `{N-x <= available_regs}` and `{x <= available_smem}`. For example, when `x > available_smem`, we should stop moving buffers from registers to shared memory. The loop continues until no more buffers can be moved to shared memory. Otherwise, we may encounter invalid states, such as `{regs_buffer > available_regs, smem_buffer > available_smem}` or {regs_buffer < available_regs, smem_buffer > available_smem}.

In summary, this process achieves two goals:
(1) Ensures that as many register buffers as possible are moved to shared memory.
(2) Ensures there are enough registers and shared memory to store the buffers.

Also, as you suggested, we may start with `{regs_buffer = 0, smem_buffer = N}`, and move from smem to registers. This seems simplier, let me give it a shot.